aten/src/ATen/native/cuda/AtomicOps.cuh

```diff
@@ -0,0 +1,344 @@
+    __attribute__((enable_if(true, "")))
+#pragma GCC diagnostic pop
+#endif
+  {
+    AtomicFPOp<double>()(address, val,
+                         [](double val, unsigned long long int assumed) {
+                           return __double_as_longlong(val + __longlong_as_double(assumed));
+                         });
+  }
+#endif
+
+  // default double and float types call CUDA atomicAdd.
+  template <typename T>
+  inline __device__ void operator() (T * address, T val) {
+    atomicAdd(address, val);
+  }
+
+  template <typename T>
+  inline __device__ void operator() (c10::complex<T> * address, c10::complex<T> val) {
+    operator()(&address->real_, val.real_);
+    operator()(&address->imag_, val.imag_);
+  }
+};
+
+template <>
+struct gpuAtomic<mul_op> {
+  inline __device__ void operator() (uint8_t * address, uint8_t val) {
+    AtomicIntegerop<uint8_t, sizeof(uint8_t)>()(address, val,
+                                                [](uint8_t val, uint32_t old_byte) {
+                                                  return static_cast<uint8_t>(THCNumerics<uint8_t>::mul(val, old_byte));
+                                                });
+  }
+
+  inline __device__ void operator() (int8_t * address, int8_t val) {
+    AtomicIntegerop<int8_t, sizeof(int8_t)>()(address, val,
+                                              [](int8_t val, uint32_t old_byte) {
+                                                return static_cast<int8_t>(THCNumerics<int8_t>::mul(val, old_byte));
+                                              });
+  }
+
+  inline __device__ void operator() (int16_t * address, int16_t val) {
+    AtomicIntegerop<int16_t, sizeof(int16_t)>()(address, val,
+                                                [](int16_t val, uint32_t old_byte) {
+                                                  return static_cast<uint16_t>(THCNumerics<int16_t>::mul(val, old_byte));
+                                                });    
+  }
+
+  inline __device__ void operator() (int32_t * address, int32_t val) {
+    AtomicIntegerop<int32_t, sizeof(int32_t)>()(address, val,
+                                                [](int32_t val, uint32_t old) {
+                                                  return val * (int32_t)old;
+                                                });
+  }
+
+  inline __device__ void operator() (int64_t * address, int64_t val) {
+    AtomicIntegerop<int64_t, sizeof(int64_t)>()(address, val,
+                                                [](int64_t val, unsigned long long old) {
+                                                  return val * (int64_t)old;
+                                                });    
+  }
+
+  inline __device__ at::Half operator() (at::Half * address, at::Half val) {
+    return AtomicFPOp<at::Half>()(address, val,
+                                  [](at::Half bsum, at::Half val) {
+                                    return THCNumerics<at::Half>::add(bsum, val);
```

Comment:
looks like copy paste, should be mul instead of add. Strange that it is not caught by tests.

Reply:
Fixed.